clippy_lints/src/unused_peekable.rs

```diff
@@ -0,0 +1,208 @@
+                            //
+                            // If the Peekable is passed to a function, stop
+                            ExprKind::Call(_, args) => {
+                                if let Some(func_did) = fn_def_id(self.cx, expr)
+                                    && let Ok(into_iter_did) = self
+                                        .cx
+                                        .tcx
+                                        .lang_items()
+                                        .require(LangItem::IntoIterIntoIter)
+                                    && func_did == into_iter_did
+                                {
+                                    // Probably a for loop desugar, stop searching
+                                    return;
+                                }
+
+                                for arg in args.iter().map(|arg| peel_ref_operators(self.cx, arg)) {
+                                    if let ExprKind::Path(_) = arg.kind
+                                        && let Some(ty) = self
+                                            .cx
+                                            .typeck_results()
+                                            .expr_ty_opt(arg)
+                                            .map(Ty::peel_refs)
+                                        && match_type(self.cx, ty, &paths::PEEKABLE)
+                                    {
+                                        self.found_peek_call = true;
+                                        return;
+                                    }
+                                }
+                            },
+                            // Peekable::peek()
+                            ExprKind::MethodCall(PathSegment { ident: method_name, .. }, [arg, ..], _) => {
+                                let arg = peel_ref_operators(self.cx, arg);
+                                let method_name = method_name.name.as_str();
+
+                                if (method_name == "peek"
+                                    || method_name == "peek_mut"
+                                    || method_name == "next_if"
+                                    || method_name == "next_if_eq")
+                                    && let ExprKind::Path(_) = arg.kind
+                                    && let Some(ty) = self.cx.typeck_results().expr_ty_opt(arg).map(Ty::peel_refs)
+                                    && match_type(self.cx, ty, &paths::PEEKABLE)
+                                {
+                                    self.found_peek_call = true;
+                                    return;
+                                }
+                            },
+                            // Don't bother if moved into a struct
+                            ExprKind::Struct(..) => {
+                                self.found_peek_call = true;
+                                return;
+                            },
+                            _ => {},
```

Comment:
I think this should be flipped, something like

```suggestion
                            ExprKind::AddrOf(..) | ExprKind::Unary(..) | ExprKind::DropTemps(_) => {},
                            _ => {
                                self.found_peek_call = true;
                                return;
                            },
```

So we're sure we're catching all the usages, e.g. tuples is missed here currently I think. But also weirder stuff like index assignments, returns/breaks, etc.

May need to add `return` to the end of both `Call/MethodCall` arms to terminate them earlier for this to work